NBT Locking Support

pupnewfster opened the pull request:
Summary:
- I made the locks map private as to ensure it gets added to properly. addToLock(ItemStack, RequirementHolder) can be used to directly add to locks if need be. This was necessary to better support wildcard metadata values and make NBT for items a bit easier.
- The config once again supports metadata values for items (bone meal works by default now, I also added support for the config to accept wildcard metadata values '*').
- Default values for weapons, tools, and armor now have the wildcard metadata by default so that once a tool is used it keeps on requiring the correct skill level (stops people from using a tool once and then giving it to someone without the requirements).
- While testing the NBT locks I discovered that if enforcing fake players was true, then players could use attack mobs without meeting the required skill levels. (I fixed this and also included the fix in the other events even though they seemed to somehow be working already).

pupnewfster committed:
Revert "BROKEN BUILD - Started work on trying to fix CrT and NBT Issues to no great effect"

This reverts commit 6019a1a27eb0243db7affa6c6825f15fbb8609a3.

pupnewfster committed:
Revert "BROKEN BUILD - Started work on trying to fix CrT and NBT Issues to no great effect"

This reverts commit 6019a1a27eb0243db7affa6c6825f15fbb8609a3.

pupnewfster committed:
Fixes issues brought up in #44 with NBT
Fixes #52 where the config no longer was supporting metadata (bonemeal by default was broken because of this fact). Also allows * as a wildcard metadata in the config.
Added wildcard metadata values to default config values for tools and armor (Makes it so that it keeps obeying the skill locks by default instead of having to be manually specified)

pupnewfster committed:
Fixes issues brought up in #44 with NBT
Fixes #52 where the config no longer was supporting metadata (bonemeal by default was broken because of this fact). Also allows * as a wildcard metadata in the config.
Added wildcard metadata values to default config values for tools and armor (Makes it so that it keeps obeying the skill locks by default instead of having to be manually specified)

pupnewfster committed:
Fixed being able to attack and damage entities without required level
Remove duplicate fake player config check (else if !isFake was always true)

pupnewfster committed:
Fixed being able to attack and damage entities without required level
Remove duplicate fake player config check (else if !isFake was always true)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Coders-After-Dark/Reskillable?pullRequest=54) <br/>All committers have signed the CLA.

pupnewfster commented:
The dif is slightly hard to read as GitHub incorrectly thinks some lines are changes when the entire method was rewritten/deleted which I apologize for but I cannot do anything to fix. Also if this is accepted it would be nice if a new version could be released to curseforge with the fix for NBT locks without users having to wait until the next major version with all the changes you have made issues for in the tracker.

Lanse505 commented:
Mind Gisting a Script Example of NBT support working?
Also, how does the new NBT lock work?
I'm guessing your implementation only do "fuzzy" NBT if nothing is set?
Then if there is a matching NBT lock available it'll use that lock value instead?

As well as the new Config syntax for getting Metadata to work?

Also is there any changes that needs to be done in CompatSkills CrT integration to get this to work?
**CrT Integration in CompatSkills:** 
https://github.com/Coders-After-Dark/CompatSkills/blob/master/src/main/java/codersafterdark/compatskills/common/compats/crafttweaker/RequirementTweaker.java

Lanse505 commented:
Also did you double check so that the changes were done to fake player vs real player checking still makes it so if the config option is enabled for fake player gating, it's still working.

pupnewfster commented:
For initial testing I used `mods.compatskills.Requirement.addRequirement(<minecraft:diamond_sword:*>.withTag({ench:[{lvl:1 as short, id: 17 as short}]}), "reskillable:attack|20");` to test NBT locking (it was the first enchantment I got on a diamond sword). I did test some variations on it (using tinkers) but removed them from the script file when I was testing fake players and things because I had removed tinkers from my test pack.

NBT locking works by first checking if there is a lock with the same NBT and if there is not it then checks to see if there is a lock set for that item without NBT and uses it. I created the ItemInfo class so that it can find things in the map with support for wildcard metadata so that the map does not have to be looped over for each entry there is.

The "new Config syntax" follows the text in the syntax given in the config
```
# Set requirements for items in this list. Each entry is composed of the item key and the requirements
# The item key is in the simple mod:item_id format. Optionally, it can be in mod:item_id:metadata, if you want to match metadata.
```
except that it also now supports * as metadata. The reason for the change is that by default it does not currently support bonemeal because metadata support in the config was currently broken as a whole.

Compat Skill should work fine as is, given it adds to crafttweakerLocks so my changes to the main locks map won't cause it problems.

I did do some testing to make sure the changes done to fake player vs real player that it still works.

If you have any other questions feel free to ask.

Lanse505 commented:
So I actually have a question related to something we had planned for the new system that you might be able to look into?

Basically, as it stands, if you have several matching locks.
I think (haven't had time to double check) from memory that only one matching lock will be applied.
The idea we are looking into doing in the future would be that if you have several matching locks, it'd iterate through any matching locks and then take the highest result for each requirement and apply it to the itemstack.

The suggested use-case is for example with tinker gating.

Where in you'd have a Tico Pickaxe with:
Bone, Stone, and Flint.

Each material would have its own NBT lock set-up for the pickaxe.
And the final lock for the pickaxe would be equal to:
The total requirements of Bone, Stone, and Flint materials.
And if a requirement is "the same" for example:
Bone = Mining|8
Flint = Mining|9
Stone = Mining|10

Then It'd default the "Mining" requirement to the highest available requirement value so 10.

It's a bit OOS for this PR, however, if you feel like this would be something you'd want to look into it, it'd be much appreciated.

If not I'll wait for Sky to get back from his work retreat and we'll look over the PR and if everything is good code-wise we'll pr it in and release an update.

pupnewfster commented:
My system supports having multiple different NBT locks for each item, and it looks for a matching one and uses it. Is that what you are asking, or are you asking about partial NBT locks where it looks for a specific tag component and finds the closest one? I think this pull may in fact already do what you are asking (though for each tool combination you would probably need to set the lock value as in stone head, wooden tool rod etc.) However, it does support adding a general lock requirement for tinkers construct pickaxes, and then for each one adding a specific higher or even lower lock.

Edit: My suggestion for how to do this is either creating some sort of NBT closeness algorithm where it looks for the one with "some" matching tags, or potentially having Tinker's support be in CompatSkills and have it calculate and add all different combinations for a material to crafttweakerLocks (or by using the addToLock method. Both probably would need to be done because I think added modifiers to tinker's tools probably have different NBT information.

Lanse505 commented:
Okey so basically something like this:
```
mods.compatskills.Requirement.addRequirement(<tconstruct:pickaxe>.withTag({TinkerData : {Materials:["manasteel"]}}), "reskillable:mining|8", "adv|minecraft:husbandry/plant_seed");
mods.compatskills.Requirement.addRequirement(<tconstruct:pickaxe>.withTag({TinkerData : {Materials:["elementium"]}}), "reskillable:mining|9", "trait|reskillable:fossile_digger");
mods.compatskills.Requirement.addRequirement(<tconstruct:pickaxe>.withTag({TinkerData : {Materials:["terrasteel"]}}), "reskillable:mining|10", "stage|banana");
```

This would, in theory, using CrT set-up 3 locks:
Pickaxe with tag matching material Manasteel: mining 8 and advancement "plant seed".
Pickaxe with tag matching material Elementium: mining 9 and trait "fossil_digger"
Pickaxe with tag matching material Terrasteel: mining 10 and stage (gamestage) "banana"

If I now had a pickaxe that used those 3 materials.
The idea is that it'd see that "Oh so there are 3 matching locks for this itemstack".
It'd then compile the locks into "one" essentially.

So the final lock for the itemstack would be:
Mining|10, Advancement: minecraft:husbandry/plant_seed, Trait: "fossil_digger", Stage: "banana". 

If that makes more sense?

I'm currently doing some IRL stuff on my weekend off so I'm a bit scattered brained atm sorry ^^'

pupnewfster commented:
I do agree with you that this out of the scope of this pull request, but in theory this could be done. The best way to implement it would probably be to have a different CraftTweaker method to add it so that it knows it is a Tinker's Material rather than you wanting a Tinker's tool with that exact NBT (some people may want that for specific circumstances). I would probably then create a class that stores an NBTTagCompound and if it is a Tinker's Material and then when looping over if it finds an exact match for the tool it uses it (like this pull request does) otherwise if it finds a matching material it would start building lock information for the tool. If you would like I can attempt to implement this in another pull request (after this one gets accepted as it would be building off the logic this one introduces and GitHub gets confused at times with different branches that build off of each other and then get merged and what changed in them). One thing that is still unclear to me is would it be for the material used at all in the pickaxe? Or just for the pickaxe head. (AKA if you used terrasteel for the tool rod would you still have mining 10?) Also as I would create a new CraftTweaker method I probably would have it have a specific format for NBT as it would be for Tinker's Tools (for example just having the nbt of the material and/or the tool modifiers). I am not quite sure what I will make the format yet (I may just have it be extra parameters of the method), but having a custom format will probably make the compiling the "full" tool information easier.


Also on second thought about changes to CompatSkills one possible improvement would be to change this line https://github.com/Coders-After-Dark/CompatSkills/blob/master/src/main/java/codersafterdark/compatskills/common/compats/crafttweaker/RequirementTweaker.java#L32 to `LevelLockHandler.addToLock(i, h);` and then you can remove the craftTweakerLocks map all together, as it would be using a more API typed method rather than adding it to a map that will iterate over the entries and call addToLock.

Lanse505 commented:
Will implement the CrT compat change here in a few.
That sounds good, as I said we probably won't pull the PR until both I and Sky can look it over :P 
Either way thanks a lot man!

pupnewfster commented:
Ok sounds good, also looking at it when you implement the CrT compat changes and are removing craftTweakerLocks from the LeveLLockHandler, you may need to remove https://github.com/pupnewfster/Reskillable/blob/nbt/src/main/java/codersafterdark/reskillable/base/LevelLockHandler.java#L53 as well because they may end up being added before setupLocks() is called. Or if someone extends Reskillable and uses addToLock before setupLocks gets called.

SkySom commented on src/main/java/codersafterdark/reskillable/base/LevelLockHandler.java:
```diff
@@ -57,57 +55,92 @@ public static void setupLocks() {
+
+        @Override
+        public boolean equals(Object o) {
+            if (o == this)
```
Always do ```if (o == this) {```

pupnewfster replied:
I will implement that change, however I am curious as to why you say to always do it. Is there a specific reason or is it just a syntax that you prefer having brackets for single line if statements or just this if statement in general for equals methods.

SkySom replied:
General coding style preferences for all things.

pupnewfster commented:
@Lanse505 While I am making the change(s) that Sky requested would you like me to do the change I mentioned of removing the craftTweakerLocks map? If I do, the change I mentioned above of how to add to the locks from CompatSkills will need to be done, when/if this gets accepted.

Lanse505 commented:
@pupnewfster sure sounds good.

pupnewfster committed:
Addresses requested changes
Renamed addToLock to addLock as it better describes what it does
Properly handle if a new lock is added once tooltips are being viewed
Removed craftTweakerLocks in favor of having addLock be called from CompatSkills

pupnewfster committed:
Merge remote-tracking branch 'pupnewfster/nbt' into nbt

# Conflicts:
#	src/main/java/codersafterdark/reskillable/base/LevelLockHandler.java

pupnewfster commented:
Git is being weird so I am going to try and fix it thinking a lot more files were changed than were and then I will say it is done.

pupnewfster commented:
Ok, I believe this is ready to be reviewed once more and then merged.

Edit: Once this is merged for continued compatibility with CompatSkills https://github.com/Coders-After-Dark/CompatSkills/blob/master/src/main/java/codersafterdark/compatskills/common/compats/crafttweaker/RequirementTweaker.java#L32 will need to be changed to `LevelLockHandler.addLock(i, h);`

SkySom commented on src/main/java/codersafterdark/reskillable/base/LevelLockHandler.java:
```diff
@@ -182,9 +229,11 @@ public static void rightClickBlock(RightClickBlock event) {
                 event.setUseBlock(Result.DENY);
                 event.setUseItem(player.isSneaking() ? Result.DEFAULT : Result.DENY);
+                if (!isFake(player))
+                    tellPlayer(player, stack, MessageLockedItem.MSG_BLOCK_USE_LOCKED);
```
Still missed a few!

SkySom commented on src/main/java/codersafterdark/reskillable/base/LevelLockHandler.java:
```diff
@@ -202,8 +251,10 @@ public static void onBlockBreak(BreakEvent event) {
+            if (!player.isCreative() && !canPlayerUseItem(player, stack)) {
                 event.setCanceled(true);
+                if (!isFake(player))
+                    tellPlayer(player, stack, MessageLockedItem.MSG_BLOCK_BREAK_LOCKED);
```
Also here

Lanse505 commented:
If @SkySom gives it the all clear by tonight Ill merge it.

Gonna be doing some fix-ups once i get home.

Lanse505 commented:
@pupnewfster will these config changes break existing configs for Reskillable?

pupnewfster commented:
No existing configs will be fine, in fact bonemeal will start working again (it currently does not properly get locked). The only thing that is different is that new configs will by default support damaged tools whereas the current config does not support them at all.

SkySom commented:
Totally forgot to click submit. But as soon as you get those changed, and @Lanse505 does a test, I'm happy to merge.

pupnewfster committed:
Missing Brackets

pupnewfster commented:
Fixed, @Lanse505 do not forget to change that CompatSkills line to use addLock.

Lanse505 commented:
@pupnewfster 
I might have found an issue with your NBT handling.

I've got two CrT added locks:
```
mods.compatskills.Requirement.addRequirement(<minecraft:wooden_pickaxe:*>, "reskillable:mining|5");
mods.compatskills.Requirement.addRequirement(<minecraft:wooden_pickaxe:*>.withTag({ench: [{lvl: 1 as short, id: 33 as short}]}), "reskillable:mining|9");
```

The first one is just a Wooden Pickaxe with any metadata
The second one is a Wooden Pickaxe with any metadata but enchanted with Silk Touch.

The issue is that the lock does not change properly it seems.
https://i.imgur.com/4Ki7XGb.png
https://i.imgur.com/S0X07ui.png


EDIT
This might actually be an advocacy point for the request for the gate combining we discussed earlier.
Because I think this might have to do with an item stacks having several matching locks.

pupnewfster commented:
That is very interesting and I believe I see why that is the case. My guess is that you combined it with silk touch in an anvil? If that is the case then this is actually working CORRECTLY as apparently when you combine a tool with an enchanted book in an anvil it also adds a RepairCost tag: `{ench: [{lvl: 1 as short, id: 33 as short}], RepairCost: 1}`. This displays as a different NBT tag so it properly uses the fallback one that does not have any NBT specified.

Lanse505 commented:
Yea that seems to be the issue, reloading atm to double check.

Lanse505 commented:
Yes, that fixed it, so everything seems to work just fine.

@pupnewfster 
Would you be interested in possibly doing the lock combining PR I suggested earlier?
Ofc that would be a separate PR.

Either way, I'm gonna go ahead and merge this PR and throw together a build package to throw on CF later today.

pupnewfster commented:
Ok that is great. I would be willing to work on a lock combining PR as you suggested earlier (may take a bit of time especially with testing but I can probably start working on it sometime this week.) When I do so I may also look into making it so that it can at times ignore things like the RepairCost tag by having a different more API oriented method that lets people add for partial locks.

Once again (I figure that you already did so given you were able to test) but do not forget to also upload a new version of CompatSkills when you are uploading this as it requires them being added with addLock instead of adding them to craftTweakerLocks which no longer exists.

Lanse505 commented:
@pupnewfster 
Feel free to join our discord btw if you wanna talk more about the mod or just hang out.
Discord link should be found on our CF page.

Already fixed the CompatSkills stuff on my side :P 
Would be hard to test CrT NBT compat without that change against your build otherwise ;)
